core/src/impl/Kokkos_ViewHooks.hpp

```diff
@@ -0,0 +1,475 @@
+// the internal view must be a reference for update_view
+// to work
+template <typename View>
+class ViewHolderRef : public ViewHolderBase {
+ public:
+  using view_type    = View;
+  using memory_space = typename view_type::memory_space;
+  using view_hook_deepcopy =
+      Kokkos::Experimental::Impl::ViewHookDeepCopy<view_type>;
+  using view_hook_copyview =
+      Kokkos::Experimental::Impl::ViewHookCopyView<view_type>;
+  using view_hook_update =
+      Kokkos::Experimental::Impl::ViewHookUpdate<view_type>;
+  explicit ViewHolderRef(view_type &view) : m_view(view) {}
+  virtual ~ViewHolderRef() = default;
+
+  size_t span() const override { return m_view.span(); }
+  bool span_is_contiguous() const override {
+    return m_view.span_is_contiguous();
+  }
+  void *data() const override { return (void *)m_view.data(); };
+
+  void *rec_ptr() const override {
+    return (void *)m_view.impl_track().template get_record<memory_space>();
+  }
+
+  ViewHolderRef *clone() const override { return new ViewHolderRef(*this); }
+
+  std::string label() const noexcept override { return m_view.label(); }
+  Kokkos::Profiling::SpaceHandle space_handle() const noexcept override {
+    return Kokkos::Profiling::make_space_handle(memory_space::name());
+  }
+  virtual bool is_data_type_const() const noexcept override {
+    return (std::is_const<typename view_type::value_type>::value);
+  }
+  size_t data_type_size() const noexcept override {
+    return sizeof(typename View::value_type);
+  }
+  bool is_hostspace() const noexcept override {
+    return std::is_same<memory_space, HostSpace>::value;
+  }
+
+  void deep_copy_to_buffer(unsigned char *buff) override {
+    view_hook_deepcopy::deep_copy(buff, m_view);
+  }
+
+  void deep_copy_from_buffer(unsigned char *buff) override {
+    view_hook_deepcopy::deep_copy(m_view, buff);
+  }
+
+  void copy_view_to_buffer(unsigned char *buff) override {
+    view_hook_copyview::copy_view(buff, m_view);
+  }
+
+  void copy_view_from_buffer(unsigned char *buff) override {
+    view_hook_copyview::copy_view(m_view, buff);
+  }
+
+  void update_view(const void *src_rec) override {
+    view_hook_update::update_view(m_view, src_rec);
+  }
+
+ private:
+  view_type &m_view;
+};
+
+struct ViewHookCallerBase {
+  virtual bool apply_to_copy_construct()                   = 0;
+  virtual void do_call(ViewHolderBase &)                   = 0;
+  virtual void do_call(ViewHolderBase &, ViewHolderBase &) = 0;
+  virtual ~ViewHookCallerBase();
+};
+
+template <class F, class ConstF = void>
+struct ViewHookCaller;
+
+template <class F>
+struct ViewHookCaller<F, void> : public ViewHookCallerBase {
+  F fun;
+  inline ViewHookCaller(F &_f) : fun(_f) {}
+  inline virtual ~ViewHookCaller() {}
+  virtual bool apply_to_copy_construct() { return false; }
+  // this is a no-op for the single case
+  virtual void do_call(ViewHolderBase &, ViewHolderBase &) {}
+  virtual void do_call(ViewHolderBase &vh) { fun(vh); }
```

Comment:
Please use `final` or `override` for this (depending on whether this is supposed to be overrideable). Don't use the keyword `virtual` for overrides.

Reply:
done in 60b42a7d87599a9d2c2bfe913cb811b38a3b6f9f